### PR TITLE
Swift extension for RTDB

### PR DESCRIPTION
**Pull request by mortenbekditlevsen**
This is a WIP. 
Tests are in place, but API still needs discussing.

### Discussion points

* I have designed the `DataSnapshot+ReadEncodable` extension to return an actual T or throw. This is different than from how it is done in the corresponding Firestore swift wrapper. My argument for doing it is that the version presented in Firestore basically wraps a true value in two levels of monads. First the implicit wrapping of having a throwing function and secondly the `Optional`. At the call site this means that you basically have to unwrap twice. The sample code presented here: 
https://github.com/firebase/quickstart-ios/commit/a42333bdae2af4808428255805d8a5723fd52302
demonstrates this issue. You iterate over documents in a collection, so you basically know they exist, but still you have to unwrap the optional - causing the developer of the sample code to insert a fatalError.
I know that the opposite can be the case too - where you actually don't know if a value exists or not. Here my preference is to catch the error corresponding to en empty snapshot. But it's naturally a matter of taste - and any RxSwift or Combine wrappers would probably deal with the situation no matter how it was implemented.

* I throw a `Database.EncodingError.internalError` in case `ServerValue.timestamp()` cannot be case to `[String: String]`.  Might there be other preferences for this?

* I modelled `ServerTimestamp` and `ServerIncrement` property wrappers, but I have second thoughts about the API. Basically they both hide two cases of an enum as an `Optional`. This makes it impossible to model actual optional values of these types. Since `ServerIncrement` has a parameter, you furthermore need to access the projected value in order to set the increment.
I made it so that `ServerIncrement` can be used directly as an enum too, and I thinkt that the API is much more true to what is being modelled.

* Encoding the `ServerTimestamp` required a cast and encoding the `ServerIncrement` is not very elegant since it re-creates the code that is performed in the `ServerValue.increment(_:)` function. Also it will fail if the type conforming to `Numeric` and `Codable` ends up not encoding as an `NSNumber`...

### Testing

  * TODO

### API Changes

  * TODO

### CODE REVIEW

**Comment by ryanwilson** [comment]
Thank you very much for the PR! Will take a look and add comments as I go.

**Comment by mortenbekditlevsen** [comment]
I would like to add api for setting default options for Database.Encoder and Database.Decoder - like the keyCodingStrategy. Would you consider such an API to be nice?

**Comment by mortenbekditlevsen** [comment]
I added API for setting default encoder and decoder. It is used as follows:
```
Database.Encoder.defaultEncoder = {
  let encoder = Database.Encoder()
  encoder.keyEncodingStrategy = .convertToSnakeCase
  return encoder
}
```
and similarly for the decoder.
Is this API understandible?
Should the static var have a different name? Perhaps something including `factory` or `creator`?

**Comment by mortenbekditlevsen** [comment]
By the way: I cannot make the linting check fail locally. I am using 10.0.1 of clang-format like the gh action. Any insights that could help me fix the linter issues myself?

**Comment by paulb777** [comment]
Actually these errors are from `swiftformat` not `clang-format`. Sorry about the confusing log.

See https://github.com/firebase/firebase-ios-sdk#code-formatting, https://github.com/firebase/firebase-ios-sdk/blob/master/scripts/setup_check.sh#L41, and https://github.com/firebase/firebase-ios-sdk/blob/master/scripts/style.sh#L184

**Comment by mortenbekditlevsen** [comment]
Thank you @paulb777 - and apologies for not actually having a look at the script instead of just looking at the output.

The output from mint was swallowed by the script. I first had mint installed and swiftformat installed, but not through mint. Then after installing swiftformat through mint I did it in /usr/local, so it couldn't be found with the MINT_PATH set by the script. 
Now it's all good - thanks. :-)

**Comment by mortenbekditlevsen** [comment]
The Tests/third_party/EncoderTests.swift is a copy of TestJSONEncoder.swift from apple/swift.
I replaced all 'expect...' with similar XCTest APIs and all encoded JSON with their similar structure.
Finally the _testRoundTrip function had to be made generic and change implementation slightly in order to handle equality checks.
I couldn't make the `testEncodingConflictedTypeNestedContainersWithTheSameTopLevelKey` work since it tests functionality that ends up in a precondition. Perhaps the apple/swift versions of the `expect...` functions can somehow magically handle precondition failures... I haven't tested that.

With custom testing of the `ServerTimestamp` and `ServerIncrement` as well as testing inspired by both the Firestore swift apis and the original `JSONEncoder` tests from apple/swift, I think that testing is in an ok place.

I am uncertain about the naming of the two test files `EncoderTests.swift` and `RTDBEncoderTests.swift` - they are not very descriptive, and it isn't obvious why they are named differently.
I am also uncertain about whether the `EncoderTests.swift` need another copy of the `LICENSE` and `METADATA` files. Please advise. :-)

So now with testing in place I am looking forward to discussing the APIs. ;-)

**Comment by mortenbekditlevsen** [comment]
I could imagine that you would like some renaming of internal types in the `RTDBEncoder.swift`. I can try and go through it and rename all occurrences of JSON to something like 'structure' or similar.

Let me know what you think.

**Comment by mortenbekditlevsen** [comment]
With this latest commit I have made it possible to decode `Optional` types directly through the API, meaning that you can more cleanly express your expectations about values in the database:

```
  public func data<T: Decodable>(as type: T.Type,
                                 decoder: Database.Decoder = Database.Decoder
                                   .defaultDecoder()) throws -> T {
    try decoder.decode(T.self, from: value ?? NSNull())
  }
```
By decoding `value ?? NSNull()` - in case `T` is not an optional and `value` is missing, then the decoding will fail with a `DecodingError.valueNotFound`.
But if `T` is an optional, then `NSNull()` parses as nil.

This improves the API to let you express:
```
let user = snapshot.data(as: User.self)
```
(for instance if you are dealing with 'childAdded' events where you know that the values are present.)
or 
```
let user = snapshot.data(as: User?.self)
```
(for instance if you read a .value event from a location where a user may or may not be represented)

I think that makes for a pretty flexible API where you can quite clearly express you expectations.

**Review by ryanwilson — COMMENTED** [review]
Thanks again so much for this - this is quite a large PR and so I think it's best to break it down into a few steps so we can make progress on it on both sides.

Any public API we ship with Firebase requires API approval from a group of folks with a proposal doc, I can create the proposal doc once we think the API is in a good state. Thanks for your recommendations and thorough explanations on why you made those API decisions - that will be helpful in evaluating the right API to choose!

1. Let's compare this to the existing Firestore API to see where things line up and where they diverge. 
  - There may be some situations where we could update the Firestore API, and some where we make DB consistent. Regardless, I think we'll want as much consistency as possible between the two in both API surface and implementation where it makes sense
  - After this is done, I'll come back with some recommendations on what we should do
2. Create the API proposal once that's been solidified
3. Look at the implementations of the APIs and do a full pass, looking for best practices and consistency with our existing Swift code (which we can also update if there improvements to be made, in a separate PR of course)
4. Continue with the iteration back and forth as necessary until it's ready to land
5. Merge the code into `master`
6. Create a Podspec for this to prepare for publishing (we can do this as well)
7. Publish the Swift Package (at the next available release) and push the CocoaPod.
8. Celebrate! 🥳 

The first one will take the longest by far I think, just taking into account how long it took to get the Firestore API to its current state. I'll try to do small pieces at a time so we can make progress as we go instead of coming back with the full API bits. 

Me or someone on the Firestore or DB team will start making comments on specific API surfaces and we'll continue from there.

Thanks again! 😄

File: FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift
Comment (schmidt-sebastian) [reply FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift:42]: The DatabaseReference is the same reference from the callsite, so it exists even if the call itself failed. This is the pattern on Android and iOS, but it is notably not used on Web. While I would personally not use this callback style in a new product (and we opted not to for Firestore), I think we should follow the honor the precedent here and continue to use the existing callback signature.

File: FirebaseDatabaseSwift/Sources/Codable/CodableErrors.swift
Comment (mortenbekditlevsen) [reply FirebaseDatabaseSwift/Sources/Codable/CodableErrors.swift]: Thanks, fixed.

File: FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift
Comment (mortenbekditlevsen) [reply FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift:42]: For a more 'swifty' API, I would prefer a `Result<Void, Error>` completion handler, so if the `DatabaseReference` can be skipped I would vote for that change. My gut feeling about the intention of this task is not to provide APIs in the same style, but rather design them to fit better with expectations of swift users. But this is just my own take. :-)

With a future (pun somewhat intended) Combine-wrapper API, this would probably translate to returning a (discardableResult) `Future<Void, Error>`. I don't know how the different versions might be disambiguated. This could probably be done with three different overloads - one that does not take a completion handler, one that does and one that returns the `Future`.

Comment (schmidt-sebastian) [reply FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift:42]: Wouldn't dropping the DatabaseReference make this call inconsistent with the other APIs in this repo? These APIs are already in our customer's hands and we just missed the window to make breaking changes. There is definitely something to be said about following Swift style, but we should probably also look at other APIs in our SDK for precedent.

File: FirebaseDatabaseSwift/Sources/Codable/ServerTimestamp.swift
Comment (ryanwilson) [inline FirebaseDatabaseSwift/Sources/Codable/ServerTimestamp.swift:68]: I feel pretty safe to force unwrap here instead of attempting to cast, the implementation hasn't changed for many years and would likely be a massive breaking change if so. @schmidt-sebastian any arguments there?

Comment (schmidt-sebastian) [reply FirebaseDatabaseSwift/Sources/Codable/ServerTimestamp.swift:68]: As discussed in our meeting today, let's see if we can annotate `ServerValue.timestamp()` correctly and remove this cast. This may allow us to remove all error handling in this PR.

Comment (mortenbekditlevsen) [reply FirebaseDatabaseSwift/Sources/Codable/ServerTimestamp.swift]: I changed the code to use a force cast and removed the custom error type. If the `ServerValue.timestamp()` is annotated in the future, then the force cast will give a compiler warning and could be removed.

File: FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift
Comment (mortenbekditlevsen) [reply FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift:42]: I'd say that the ongoing work in #6940 suggests that the `Result` type might be a good candidate to use here too?

Comment (schmidt-sebastian) [reply FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift:42]: We can use `Result` here.

**Comment by mortenbekditlevsen** [comment]
Hi @ryanwilson ,
Apologies for bumping this PR thread. Is there anything I can do to help progress here?

I started small experiments with Combine in #7003 inspired by efforts for adding Combine support for Auth.

**Comment by mortenbekditlevsen** [comment]
With all the recent developments with `async/await` (there's even an experimental build of a tool chain with support for trying it out), perhaps all apis involving `Result` based completion handlers ought to be postponed until they can be modeled using async/await? At least that will save a few overloads of the same functionality.
I'd recommend the same for any `Future` based apis in the efforts to add `Combine` wrappers.

**Comment by ryanwilson** [comment]
> Hi @ryanwilson ,
> Apologies for bumping this PR thread. Is there anything I can do to help progress here?
> 
> I started small experiments with Combine in #7003 inspired by efforts for adding Combine support for Auth.

Hi! No need to apologize, I should be the one apologizing for not getting back to this soon enough. It's been quite busy with the ARM Macs and App Store changes, but things are settling down now so I can dedicate some time to this (although we're rapidly approaching the holidays and folks won't be in the office, but I'll try to make some progress regardless).

I've had a few discussions about the API with team members but haven't made near enough progress as I'd like, so I need to get a better comparison vs Firestore ready and discuss the tradeoffs and differences. A vast majority of the API will be straightforward but the 2-3 things that are different and have many edge cases will take a significant amount of the total time (based on my experience with the Timestamp APIs in Firestore).

>With all the recent developments with async/await (there's even an experimental build of a tool chain with support for trying it out https://forums.swift.org/t/se-0296-async-await/42605/108), perhaps all apis involving Result based completion handlers ought to be postponed until they can be modeled using async/await? At least that will save a few overloads of the same functionality.
I'd recommend the same for any Future based apis in the efforts to add Combine wrappers (#6976).

You bring up a good point with async/await and Result types, I had a similar feeling. I'll take a look at some of the progress and see what makes sense to add or put on hold. At the very least I think this means we should be packaging those APIs separately so folks can still take advantage of it before async/await ships and becomes more standard, then we can deprecate those packages as needed.

Thanks for your continued patience here, hope to have some more progress updates for you soon!

**Comment by mortenbekditlevsen** [comment]
Hi @ryanwilson ,
Right, the `ServerValue` APIs are the ones I am most curious about too! My gut feeling is that the property wrapper APIs are not the best match for that purpose. 

The wrapper I have made for server increments can be used directly as a generic enum instead of as a property wrapper and I think that API is a more exact representation of what is going on. 

For increments that can take on many values I think that the issue becomes very explicit (hiding values behind a `nil`), but I actually think that the server side time stamp wrapper 'suffers' from the same issue... For instance with the property wrapper you can't model an actual optional server side time stamp. 
But I am sure that you have discussed this at length with regards to the similar API for Firestore, so I am curious to hear any thoughts about that. :-)

With regards to async await, some APIs are about to get three different flavors: a `Result` based completion handler flavor, the `Combine` `Future` flavor and then eventually the async flavor too. 

Since the APIs that I have been adding are just ergonomics and conveniences, my personal opinion is that it would be ok to require a specific version of Swift for the overlays, and then only include the `async` flavor (given that async await lands sometimes in the spring).

**Comment by mortenbekditlevsen** [comment]
Hi @ryanwilson ,
Would it help in any way if I reduced the size of the PR by for instance moving the ServerValue property wrappers to a separate PR?

**Comment by mortenbekditlevsen** [comment]
Hi @ryanwilson ,
Sorry to be pestering you about this.
Is there any updates about the review?
I could split up the PR into smaller bits of that is helpful for you.
Please let me know if there is anything at all I can do to assist.

**Comment by ryanwilson** [comment]
Hi @mortenbekditlevsen - sincere apologies for the delay and radio silence on this. It's far below the bar for how we want any contributions to be handled.

I blocked off some time for myself today and tomorrow to finally make progress on the API front, I'll reply back end of day tomorrow hopefully with a better grasp and some concrete suggestions, but at the very least I'll make sure I reply back and am active until this lands.

Thanks for your patience here, not a pester at all. We sincerely appreciate the contribution and I'm looking forward to getting this landed!

**Review by ryanwilson — COMMENTED** [review]
Still need to review the decoder/encoder combo and a few other small things, but some feedback for now!

File: FirebaseDatabaseSwift/Sources/Codable/DataSnapshot+ReadDecodable.swift
Comment (mortenbekditlevsen) [reply FirebaseDatabaseSwift/Sources/Codable/DataSnapshot+ReadDecodable.swift:42]: Hi Ryan,
I'll naturally go with your recommendations, but I just have a small comment to your discussion:
There's no need for using `snapshot.exists`. The user is in full control about whether or not the expect existance/optionality.
For instance, if you have a snapshot of some `Entity` that may or may not exist, then you can write:
```
snapshot.data(as: Entity?.self)
```
While if you have some `Configuration` that must exist (ie. you consider it an error for it to be missing), then you can do:
```
snapshot.data(as: Configuration.self)
```
And the final example where you have snapshots from `childAdded`. Here you know for a fact that data can exist and can parse it without optionality.

So my main point is that I believe that the user has more power in the choice of adding optionality to the generic parameter themselves rather than have it 'forced'. And I don't see the need for calling `snapshot.exists` - you just imply optionality when decoding.

File: FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift
Comment (mortenbekditlevsen) [reply FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift:53]: Certainly - this is just a matter of taste. I kind of like the 'shape' of always using `Result` for completion handlers since it directly maps to my mental model of `Future`s.

I'll make the suggested change.

File: FirebaseDatabaseSwift/Sources/Codable/DataSnapshot+ReadDecodable.swift
Comment (mortenbekditlevsen) [reply FirebaseDatabaseSwift/Sources/Codable/DataSnapshot+ReadDecodable.swift:42]: Another thought I just had:
When adding the `Combine` API to get some `Publisher` from a `DatabaseReference`, I would think that the API was similar to `data(as:)`. Would you for that API also expect that '.decodedEntityPublisher(as: Configuration.self)' would give a publisher of optional `Configuration` rather than allowing the user to specify the optionality in the generic parameter here too?

File: FirebaseDatabaseSwift/Sources/Codable/ServerIncrement.swift
Comment (mortenbekditlevsen) [reply FirebaseDatabaseSwift/Sources/Codable/ServerIncrement.swift:19]: No worries - I'll just remove that.
One suggestion could be to remove the `@propertyWrapper` and still keep the enum around. That could still give value to users of the API.
But it's of course also something that can be added in user-space too.

File: FirebaseDatabaseSwift/Sources/Codable/DataSnapshot+ReadDecodable.swift
Comment (mortenbekditlevsen) [reply FirebaseDatabaseSwift/Sources/Codable/DataSnapshot+ReadDecodable.swift:42]: Is there any possibility that my comments here may be reconsidered? :-)
I implemented the other two suggested changes.

**Comment by mortenbekditlevsen** [comment]
Hi @ryanwilson ,
I know that you're very busy but I'll just give this thread a small bump. :-)
I updated the code for two of your three comments - and I was hoping that you might discuss the third point internally again. 
Sincerely,
/Morten

**Comment by ryanwilson** [comment]
@mortenbekditlevsen thank you for the very kind reminder, taking a look this week and I'll do my best to get things to a state where we can merge. Thanks again for your immense patience on this!

File: FirebaseDatabaseSwift/Sources/third_party/RTDBEncoder/RTDBEncoder.swift
Comment (ryanwilson) [inline FirebaseDatabaseSwift/Sources/third_party/RTDBEncoder/RTDBEncoder.swift:1]: For future readers - I diffed this with the copy from the Swift lib and there were no significant changes.

File: FirebaseDatabaseSwift/Tests/Codable/RTDBEncoderTests.swift
Comment (ryanwilson) [inline FirebaseDatabaseSwift/Tests/Codable/RTDBEncoderTests.swift:34]: As mentioned above with the default encoders, I think instantiating a new encoder for each test would be more clear. Then we can also test custom snakeCase encoding there as well.

File: FirebaseDatabaseSwift/Tests/third_party/EncoderTests.swift
Comment (ryanwilson) [inline FirebaseDatabaseSwift/Tests/third_party/EncoderTests.swift:1]: For future readers: diffed the original copy and this one and there were no major changes.

File: FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift
Comment (ryanwilson) [reply FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift:24]: Looked at it further and I'm not convinced wee need the `defaultEncoder` - was there a specific reason you used this vs using `Database.Encoder()` as the default value below?

**Review by ryanwilson — COMMENTED** [review]
Thanks again - a few questions left but we're super close to getting this landed. I'm happy to set up CI for it afterwards, unless you'd like to see how our GitHub actions workflows go!

File: FirebaseDatabaseSwift/Sources/Codable/DataSnapshot+ReadDecodable.swift
Comment (mortenbekditlevsen) [reply FirebaseDatabaseSwift/Sources/Codable/DataSnapshot+ReadDecodable.swift:42]: Hi @ryanwilson ,
I haven't been able to create mock instances of `DataSnapshot` in Swift. The last time I checked I couldn't find any pointers in the codebase as to how I might do it.
Do you have any suggestions here?

File: FirebaseDatabaseSwift/Sources/Codable/ServerTimestamp.swift
Comment (mortenbekditlevsen) [reply FirebaseDatabaseSwift/Sources/Codable/ServerTimestamp.swift:138]: I think I was just inspired by similar version checks in the Firestore equivalent - I'll just remove it.
I hadn't even created a podspec for this yet - and I don't know exactly how you would like this.
I am going to copy FirebaseFirestoreSwift.podspec, use the same version number (0.4.0) and then use the description from FirebaseDatabase.podspec.

File: FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift
Comment (mortenbekditlevsen) [reply FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift:24]: I just thought it would be a nice touch for the consumer of the API to define a singleton preferred configuration. But it's no big deal - I am removing it for the next iteration. :-)

File: FirebaseDatabaseSwift.podspec
Comment (ryanwilson) [inline FirebaseDatabaseSwift.podspec:8]: Nit: This  should be `7.9.0-beta` as we have a new naming scheme.

File: FirebaseDatabaseSwift.podspec
Comment (ryanwilson) [inline FirebaseDatabaseSwift.podspec:21]: Nit: `'CocoaPods-' + s.version.to_s` now.

File: FirebaseDatabaseSwift.podspec
Comment (ryanwilson) [inline FirebaseDatabaseSwift.podspec:25]: Nit: iOS deployment target is now 10f or RTDB

File: FirebaseDatabaseSwift.podspec
Comment (ryanwilson) [inline FirebaseDatabaseSwift.podspec:26]: This is now `10.12`

File: FirebaseDatabaseSwift.podspec
Comment (ryanwilson) [inline FirebaseDatabaseSwift.podspec:30]: Nit: this line can be removed

File: FirebaseDatabaseSwift.podspec
Comment (ryanwilson) [inline FirebaseDatabaseSwift.podspec:33]: This can be removed  since it's not relevant for Swift files

File: FirebaseDatabaseSwift.podspec
Comment (ryanwilson) [inline FirebaseDatabaseSwift.podspec:38]: For the version  dependency, can you please change it to `'~> 7.8'`?

**Review by ryanwilson — COMMENTED** [review]
Thanks very  much for adding the podspec! There have been a few changes since you originally opened the PR,  I've added comments at what needs to change. I think after that I'll go through the official API review and we'll be good to publish. Thanks again for all the hard work here and I'm excited to finally get this in the repo.

**Comment by mortenbekditlevsen** [comment]
Hi @ryanwilson ,
Thanks for the comments - I updated the podspec.
Do you have any thoughts about the issues with mocking snapshots? I think I'll run into similar issues for the future PRs concerning RTDB and Combine.

File: Package.swift
Comment (ryanwilson) [inline Package.swift:48]: Ah, one last thing! I forgot that we moved to append a `-Beta` to the Swift libs in SwiftPM:

https://github.com/firebase/firebase-ios-sdk/blob/f5fbdac038ac6fa284074ea2a101f22a573bbb87/Package.swift#L86-L89

After that I'm good to merge this and make any API changes as necessary from the API review in a follow up PR. Thank you!

Comment (mortenbekditlevsen) [reply Package.swift]: Done! Looking forward to the merge. Thank you so much for all your help! :-)

**Review by ryanwilson — APPROVED** [review]
Thank you very much for everything! I'm not sure if we'll make it into the next release but it may have to wait since the API review still hasn't happened. I'll update this issue once it gets published. Congrats and thanks again for this awesome contribution!

**Comment by mortenbekditlevsen** [comment]
You're very welcome! I'm thrilled to have been able to contribute - and now I am looking forward to give a shot at the Combine efforts. :-)
And thank you for all your help! 😊

**Pull request merged**
